src/server/frontend_wayland/xdg_shell_v6.cpp

```diff
@@ -413,25 +417,43 @@ void mf::XdgSurfaceV6::set_min_size(int32_t width, int32_t height)
 
 void mf::XdgSurfaceV6::set_maximized()
 {
-    if (surface_id.as_value())
-    {
-        spec().state = mir_window_state_maximized;
-    }
-    else
-    {
-        params->state = mir_window_state_maximized;
-    }
+    set_state(mir_window_state_maximized);
 }
 
 void mf::XdgSurfaceV6::unset_maximized()
+{
+    set_state(mir_window_state_restored);
+}
+
+void mf::XdgSurfaceV6::set_fullscreen(std::experimental::optional<struct wl_resource*> const& output)
+{
+    (void)output; // TODO specify the output when setting fullscreen
+    set_state(mir_window_state_fullscreen);
+}
+
+void mf::XdgSurfaceV6::unset_fullscreen()
+{
+    set_state(mir_window_state_restored);
+}
+
+void mf::XdgSurfaceV6::set_minimized()
+{
+    set_state(mir_window_state_minimized);
+}
+
+void mf::XdgSurfaceV6::set_state(MirWindowState state)
```

Comment:
perhaps this belongs in `WlAbstractMirWindow`? Also, I would change the name to `set_state_immediately` or something else that clearly indicates these changes are not double buffered.

Reply:
Renamed.

Moving the functions to WlAbstractMirWindow is plausible, but as it stands we also need to move XdgSurfaceV6::shell. There could also be other commonality. I've left that for a future refactoring.

bors r=wmww,RAOF